ndcube/meta.py

```diff
@@ -0,0 +1,374 @@
+        super().__init__(meta.items())
+        meta_keys = meta.keys()
+
+        if comments is None:
+            self._comments = dict()
+        else:
+            comments = dict(comments)
+            if not set(comments.keys()).issubset(set(meta_keys)):
+                raise ValueError(
+                    "All comments must correspond to a value in meta under the same key.")
+            self._comments = comments
+
+        if data_shape is None:
+            self._data_shape = data_shape
+        else:
+            self._data_shape = np.asarray(data_shape, dtype=int)
+
+        if axes is None:
+            self._axes = dict()
+        else:
+            if not (isinstance(data_shape, collections.abc.Iterable) and
+                    all([isinstance(i, numbers.Integral) for i in data_shape])):
+                raise TypeError("If axes is set, data_shape must be an iterable giving "
+                                "the length of each axis of the associated cube.")
+            axes = dict(axes)
+            if not set(axes.keys()).issubset(set(meta_keys)):
+                raise ValueError(
+                    "All axes must correspond to a value in meta under the same key.")
+            self._axes = dict([(key, self._sanitize_axis_value(axis, meta[key], key))
+                               for key, axis in axes.items()])
+
+    def _sanitize_axis_value(self, axis, value, key):
+        axis_err_msg = ("Values in axes must be an integer or iterable of integers giving "
+                        f"the data axis/axes associated with the metadata.  axis = {axis}.")
+        if isinstance(axis, numbers.Integral):
+            axis = (axis,)
+        if len(axis) == 0:
+            return ValueError(axis_err_msg)
+        if self.shape is None:
+            raise TypeError("Meta instance does not have a shape so new metadata "
+                            "cannot be assigned to an axis.")
+        # Verify each entry in axes is an iterable of ints or a scalar.
+        if not (isinstance(axis, collections.abc.Iterable) and all([isinstance(i, numbers.Integral)
+                                                                    for i in axis])):
+            return ValueError(axis_err_msg)
+        axis = np.asarray(axis)
+        if _not_scalar(value):
+            axis_shape = tuple(self.shape[axis])
+            if not _is_grid_aligned(value, axis_shape) and not _is_axis_aligned(value, axis_shape):
+                raise ValueError(
+                    f"{key} must have shape {tuple(self.shape[axis])} "
+                    f"as its associated axes {axis}, ",
+                    f"or same length as number of associated axes ({len(axis)}). "
+                    f"Has shape {value.shape if hasattr(value, 'shape') else len(value)}")
+        return axis
+
+    @property
+    def comments(self):
+        return self._comments
+
+    @property
+    def axes(self):
+        return self._axes
+
+    @property
+    def shape(self):
```

Comment:
We should be consistent in how we name this, I think `data_shape` is better:
```suggestion
    def data_shape(self):
```